File: src/path.ts
/**
 * Path encoding (specification section 5).
 *
 * A path is a sequence of typed segments. There are no reserved characters and no escaping: the
 * length prefix does the work an escaping rule would otherwise do, so a nested `a.b` and a literal
 * dotted key `"a.b"` are provably distinct with no rejection rule at all (section 5.1).
 *
 * The display form lives here too, and it is NEVER hashed. `displayPath` produces it; nothing in
 * this library consumes it. Section 5.2 forbids reconstructing an encoded path by parsing a
 * display path, and there is deliberately no parser for one in this module.
 */

import { fail } from './errors.js';
import { nfc, utf8, u32be } from './bytes.js';
import { assertNoUnpairedSurrogate } from './json.js';

export interface KeySegment {
  readonly key: string;
}

export interface IndexSegment {
  readonly index: number;
}

export type PathSegment = KeySegment | IndexSegment;

/** A path, as the sequence of segments that is the only authoritative representation. */
export type Path = readonly PathSegment[];

export function isKeySegment(s: PathSegment): s is KeySegment {
  return Object.prototype.hasOwnProperty.call(s, 'key');
}

/** `2^32`. An index MUST be strictly below this (specification section 5). */
const INDEX_CEILING = 4294967296;

/**
 * Encodes a path (specification section 5).
 *
 * ```
 * encodePath(segments) =
 *     u32be(count(segments))
 *   ‖ for each segment:
 *       KEY(k)   ->  0x01 ‖ u32be(len(utf8(NFC(k)))) ‖ utf8(NFC(k))
 *       INDEX(i) ->  0x02 ‖ u32be(i)
 * ```
 *
 * The KEY length prefix counts the UTF-8 BYTES of the NORMALIZED key, not its code points and not
 * its UTF-16 code units. Those three differ for every non-ASCII key.
 */
export function encodePath(segments: Path): Uint8Array {
  const parts: Uint8Array[] = [u32be(segments.length)];
  for (const segment of segments) {
    if (isKeySegment(segment)) {
      assertNoUnpairedSurrogate(segment.key, 'a path key');
      const bytes = utf8(nfc(segment.key));
      parts.push(Uint8Array.of(0x01), u32be(bytes.length), bytes);
    } else {
      const i = segment.index;
      if (!Number.isInteger(i) || i < 0) {
        fail('index-out-of-32-bit-range', `array index ${i} is not a non-negative integer`, {
          index: i,
        });
      }
      // Section 5: indices MUST be `< 2^32`, and an implementation that cannot represent one in
      // 32 bits MUST error rather than truncate. `u32be` would wrap silently.
      if (i >= INDEX_CEILING) {
        fail('index-out-of-32-bit-range', `array index ${i} is not below 2^32`, { index: i });
      }
      parts.push(Uint8Array.of(0x02), u32be(i));
    }
  }
  return concatBytes(parts);
}

function concatBytes(parts: readonly Uint8Array[]): Uint8Array {
  let total = 0;
  for (const p of parts) {
    total += p.length;
  }
  const out = new Uint8Array(total);
  let at = 0;
  for (const p of parts) {
    out.set(p, at);
    at += p.length;
  }
  return out;
}

/**
 * The human-readable form, `a.b[0].c`.
 *
 * Display only. It is not an input to any hash, and it is not parsed back: specification section
 * 5.2 states both halves, and `corpus/README.md` records that a minimum-disclosure floor written
 * in this notation is silently unenforced because the dotted string matches no leaf a record has.
 *
 * The form is lossy by construction - a key containing `.`, `[` or `]` renders ambiguously - and
 * that is exactly why it never round-trips.
 */
export function displayPath(segments: Path): string {
  let out = '';
  for (const segment of segments) {
    if (isKeySegment(segment)) {
      out += out === '' ? segment.key : `.${segment.key}`;
    } else {
      out += `[${segment.index}]`;
    }
  }
  return out;
}

/** Unsigned byte comparison, which is the ordering specification section 9 pins. */
export function compareBytes(a: Uint8Array, b: Uint8Array): number {
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i += 1) {
    const x = a[i] as number;
    const y = b[i] as number;
    if (x !== y) {
      return x < y ? -1 : 1;
    }
  }
  if (a.length === b.length) {
    return 0;
  }
  return a.length < b.length ? -1 : 1;
}

File: rust/src/path.rs
use crate::error::{Error, Result};
use std::fmt;
use unicode_normalization::UnicodeNormalization;

/// One authoritative structured path segment.
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub enum Segment {
    Key(String),
    Index(u32),
}

/// An authoritative path. Display strings are intentionally not parseable.
#[derive(Clone, Debug, Default, PartialEq, Eq, Hash)]
pub struct Path(Vec<Segment>);

impl Path {
    #[must_use]
    pub const fn new() -> Self {
        Self(Vec::new())
    }

    #[must_use]
    pub fn from_segments(segments: Vec<Segment>) -> Self {
        Self(segments)
    }

    #[must_use]
    pub fn segments(&self) -> &[Segment] {
        &self.0
    }

    #[must_use]
    pub fn into_segments(self) -> Vec<Segment> {
        self.0
    }

    #[must_use]
    pub fn with_key(&self, key: impl Into<String>) -> Self {
        let mut segments = self.0.clone();
        segments.push(Segment::Key(key.into()));
        Self(segments)
    }

    #[must_use]
    pub fn with_index(&self, index: u32) -> Self {
        let mut segments = self.0.clone();
        segments.push(Segment::Index(index));
        Self(segments)
    }

    /// Encode the authoritative structured path.
    pub fn encode(&self) -> Result<Vec<u8>> {
        let count = u32::try_from(self.0.len())
            .map_err(|_| Error::InvalidPath("more than 2^32-1 path segments".into()))?;
        let mut out = Vec::new();
        out.extend_from_slice(&count.to_be_bytes());

        for segment in &self.0 {
            match segment {
                Segment::Key(key) => {
                    let normalized: String = key.nfc().collect();
                    let bytes = normalized.as_bytes();
                    let length = u32::try_from(bytes.len()).map_err(|_| {
                        Error::InvalidPath("UTF-8 key is longer than 2^32-1 bytes".into())
                    })?;
                    out.push(0x01);
                    out.extend_from_slice(&length.to_be_bytes());
                    out.extend_from_slice(bytes);
                }
                Segment::Index(index) => {
                    out.push(0x02);
                    out.extend_from_slice(&index.to_be_bytes());
                }
            }
        }

        Ok(out)
    }

    /// Render a non-authoritative path for diagnostics and user interfaces.
    #[must_use]
    pub fn display(&self) -> String {
        let mut rendered = String::new();
        for segment in &self.0 {
            match segment {
                Segment::Key(key) => {
                    if !rendered.is_empty() {
                        rendered.push('.');
                    }
                    rendered.push_str(key);
                }
                Segment::Index(index) => {
                    rendered.push('[');
                    rendered.push_str(&index.to_string());
                    rendered.push(']');
                }
            }
        }
        rendered
    }

    #[must_use]
    pub fn normalized(&self) -> Self {
        Self(
            self.0
                .iter()
                .map(|segment| match segment {
                    Segment::Key(key) => Segment::Key(key.nfc().collect()),
                    Segment::Index(index) => Segment::Index(*index),
                })
                .collect(),
        )
    }
}

impl From<Vec<Segment>> for Path {
    fn from(value: Vec<Segment>) -> Self {
        Self::from_segments(value)
    }
}

impl fmt::Display for Path {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.display())
    }
}
